File: web/src/lib/sharePlaces.ts
import { saveJsonFile } from "./backup";
import { PLACE_CATEGORIES, type Place, type PlaceCategory } from "../types";

/**
 * A place stripped down to what's worth handing to someone else — no id,
 * board, coordinates, visited/favorite status, or collections, since
 * those are either meaningless or private outside the sender's own board.
 * The recipient's copy gets its own id and is geocoded fresh on import.
 */
export interface SharedPlace {
  name: string;
  category: PlaceCategory;
  address: string;
  phone: string | null;
  notes: string;
  linkUrl: string | null;
}

export interface SharedPlacesPayload {
  app: "peragra";
  kind: "places";
  version: 1;
  places: SharedPlace[];
}

export function buildSharedPlacesPayload(places: Place[]): SharedPlacesPayload {
  return {
    app: "peragra",
    kind: "places",
    version: 1,
    places: places.map((p) => ({
      name: p.name,
      category: p.category,
      address: p.address,
      phone: p.phone,
      notes: p.notes,
      linkUrl: p.linkUrl,
    })),
  };
}

export function sharedPlacesToText(payload: SharedPlacesPayload): string {
  return JSON.stringify(payload, null, 2);
}

export function sharedPlacesFilename(date = new Date()): string {
  const pad = (n: number) => String(n).padStart(2, "0");
  const stamp = `${date.getFullYear()}${pad(date.getMonth() + 1)}${pad(date.getDate())}_${pad(date.getHours())}${pad(date.getMinutes())}${pad(date.getSeconds())}`;
  return `peragra_places_${stamp}.json`;
}

/** Saves selected places to a location the person picks — see
 *  saveJsonFile in backup.ts for the File System Access API/download
 *  fallback behavior this shares with the whole-app backup. */
export function saveSharedPlacesFile(payload: SharedPlacesPayload): Promise<"saved" | "cancelled" | "downloaded"> {
  return saveJsonFile(payload, sharedPlacesFilename(), "Peragra places");
}

/**
 * Parses text as a Peragra "shared places" payload — pasted from another
 * person's "Copy as text" share, or read from their "Save as file". Null
 * for anything that isn't recognizably that (not JSON, wrong shape, or a
 * places array with nothing usable in it) rather than throwing, since the
 * caller just needs a yes/no to decide how to handle the input.
 */
export function parseSharedPlacesText(text: string): SharedPlace[] | null {
  const trimmed = text.trim();
  if (!trimmed) return null;

  let data: unknown;
  try {
    data = JSON.parse(trimmed);
  } catch {
    return null;
  }
  if (typeof data !== "object" || data === null) return null;
  const obj = data as Record<string, unknown>;
  if (obj.app !== "peragra" || obj.kind !== "places" || !Array.isArray(obj.places)) return null;

  const validCategories = new Set(PLACE_CATEGORIES.map((c) => c.value));
  const places: SharedPlace[] = [];
  for (const raw of obj.places) {
    if (typeof raw !== "object" || raw === null) continue;
    const p = raw as Record<string, unknown>;
    const name = typeof p.name === "string" ? p.name.trim() : "";
    if (!name) continue;
    places.push({
      name,
      category: typeof p.category === "string" && validCategories.has(p.category as PlaceCategory)
        ? (p.category as PlaceCategory)
        : "restaurant",
      address: typeof p.address === "string" ? p.address : "",
      phone: typeof p.phone === "string" && p.phone.trim() ? p.phone : null,
      notes: typeof p.notes === "string" ? p.notes : "",
      linkUrl: typeof p.linkUrl === "string" && p.linkUrl.trim() ? p.linkUrl : null,
    });
  }
  return places.length > 0 ? places : null;
}

File: Peragra/Services/SharePlaces.swift
import Foundation

/// A place stripped down to what's worth handing to someone else — no
/// id, board, coordinates, visited/favorite status, or collections,
/// since those are either meaningless or private outside the sender's
/// own board. The recipient's copy gets its own id and is geocoded
/// fresh on import. Mirrors web/src/lib/sharePlaces.ts's SharedPlace so
/// a share copied on one platform pastes cleanly on the other.
struct SharedPlace: Codable {
    let name: String
    let category: String
    let address: String
    let phone: String?
    let notes: String
    let linkURL: String?

    enum CodingKeys: String, CodingKey {
        case name, category, address, phone, notes
        case linkURL = "linkUrl"
    }
}

struct SharedPlacesPayload: Codable {
    let app: String
    let kind: String
    let version: Int
    let places: [SharedPlace]
}

enum SharePlaces {
    static func buildPayload(from places: [Place]) -> SharedPlacesPayload {
        SharedPlacesPayload(
            app: "peragra",
            kind: "places",
            version: 1,
            places: places.map { place in
                SharedPlace(
                    name: place.name,
                    category: place.category.rawValue,
                    address: place.address,
                    phone: place.phone,
                    notes: place.notes,
                    linkURL: place.linkURLString
                )
            }
        )
    }

    static func toText(_ payload: SharedPlacesPayload) throws -> String {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        let data = try encoder.encode(payload)
        return String(decoding: data, as: UTF8.self)
    }

    static func filename(date: Date = Date()) -> String {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyyMMdd_HHmmss"
        return "peragra_places_\(formatter.string(from: date)).json"
    }

    /// Writes the payload to a fresh temp file for ShareLink/
    /// UIActivityViewController to hand to Messages, AirDrop, Files, etc.
    static func writeTempFile(_ payload: SharedPlacesPayload) -> URL? {
        guard let data = try? JSONEncoder().encode(payload) else { return nil }
        let url = FileManager.default.temporaryDirectory.appendingPathComponent(filename())
        return (try? data.write(to: url, options: .atomic)) != nil ? url : nil
    }

    /// Parses text as a Peragra "shared places" payload — pasted from
    /// another person's "Copy as Text" share, or read from their "Share
    /// as File". Nil for anything that isn't recognizably that (not
    /// JSON, wrong shape, or a places array with nothing usable in it)
    /// rather than throwing, since the caller just needs a yes/no to
    /// decide how to handle the input.
    static func parse(_ text: String) -> [SharedPlace]? {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty, let data = trimmed.data(using: .utf8) else { return nil }
        guard let payload = try? JSONDecoder().decode(SharedPlacesPayload.self, from: data) else { return nil }
        guard payload.app == "peragra", payload.kind == "places" else { return nil }
        let places = payload.places.filter { !$0.name.trimmingCharacters(in: .whitespaces).isEmpty }
        return places.isEmpty ? nil : places
    }
}
